### PR TITLE
relation: proof-linking: Add `LinkableCircuit` trait and impls

### PR DESCRIPTION
**Pull request by joeykraut**
### Purpose
The proof-linking arithmetization and gate placement is finicky and important to get right, so in order to avoid re-implementing between single and multiprover: this PR introduces a `LinkableCircuit` trait. 

This trait requires implementations to give methods for querying and setting link groups and gates. On top of the required methods, this PR builds default implementation of layout generation and application. This involves placing groups, inserting proof-linking gates, etc.

As well, I removed the "caching" behavior of the layout, instead the implementation now just modifies the group layout such that subsequent calls to `generate_layout` will return the same result by having no "unplaced" groups.

### Testing
- Unit tests pass

### CODE REVIEW

File: plonk/src/multiprover/proof_system/constraint_system.rs
Comment (akirillo) [inline plonk/src/multiprover/proof_system/constraint_system.rs:1]: Seems like we're not calling `MpcPlonkCircuit::generate_layout()` in `finalize_for_arithmetization()` yet, is this intended?

File: relation/src/proof_linking/linkable_circuit.rs [hunk 1]
```diff
@@ -1,230 +1,189 @@
...
-    /// That is, for return value `n`, the circuit will be aligned to the 2^n-th
-    /// roots of unity
-    fn current_circuit_alignment(&self) -> usize {
-        let max_alignment =
-            self.link_group_layouts.values().map(|l| l.alignment).max().unwrap_or(0);
+    /// E.g. to move public input gates to the front or plookup gates to the
+    /// back
+    fn order_gates(&mut self) -> Result<(), CircuitError> {
+        Ok(())
+    }
```
Comment (akirillo) [inline relation/src/proof_linking/linkable_circuit.rs]: IMO we should remove this default implementation, a no-op should be defined by the implementing type

**Review by akirillo — APPROVED** [review]
just a couple small nits, lgtm

File: plonk/src/multiprover/proof_system/constraint_system.rs
Comment (joeykraut) [reply plonk/src/multiprover/proof_system/constraint_system.rs:1]: Yes, the multiprover linking implementation is the next PR :)

**Pull request merged**